lib/new_relic/agent/transaction/tracing.rb

```diff
@@ -34,7 +34,12 @@ def add_segment segment, parent = nil
         end
 
         def segment_complete segment
-          @current_segment = segment.parent
+          # if parent was in another thread, remove the current_segment entry for this thread
+          if segment.parent && segment.parent.starting_thread_id != ::Thread.current.object_id
+            remove_current_segment_by_thread_id(::Thread.current.object_id)
```

Comment:
Is it possible for 2 threads to reach this code at the same time, and if so would there be a problem with the simultaneous deletion operation calls?

Reply:
Hmm, that's a great question. Since it's deleting based on the thread ID, they would definitely be trying to delete _different_ things at least, but I'm not sure how hash handles that situation. i'll look into it

Reply:
I tested this behavior out in a script, and didn't see any issues... and then i remembered the issues we had with jruby, so I checked that and jruby does have an issue with that. I'll have to add a lock for it.

Reply:
Ok. I added a lock to the methods `remove_current_segment_by_thread_id` and `set_current_segment` in Transaction, since those are the only places where we're actually modifying the hash directly.